core/src/stages/generate_pose.cpp

```diff
@@ -76,15 +107,65 @@ void GeneratePose::compute() {
 		return;
 	}
 
-	InterfaceState state(scene);
-	state.properties().set("target_pose", target_pose);
+	const auto& spawn_target_pose = [&](const geometry_msgs::PoseStamped& target_pose) {
+		InterfaceState state(scene);
+		state.properties().set("target_pose", target_pose);
 
-	SubTrajectory trajectory;
-	trajectory.setCost(0.0);
+		SubTrajectory trajectory;
+		trajectory.setCost(0.0);
 
-	rviz_marker_tools::appendFrame(trajectory.markers(), target_pose, 0.1, "pose frame");
+		rviz_marker_tools::appendFrame(trajectory.markers(), target_pose, 0.1, "pose frame");
 
-	spawn(std::move(state), std::move(trajectory));
+		spawn(std::move(state), std::move(trajectory));
+	};
+
+	spawn_target_pose(target_pose);
+
+	// If no dimension samplers were defined, we're done here
+	if (pose_dimension_samplers_.empty())
+		return;
+
+	// Use target pose as seed pose
+	geometry_msgs::PoseStamped seed_pose = target_pose;
+	tf2::Quaternion seed_q, rand_q;
+	tf2::fromMsg(seed_pose.pose.orientation, seed_q);
+	double elapsed_time = 0.0;
+	const auto start_time = std::chrono::steady_clock::now();
+	size_t spawned_solutions = 0;
+	while (elapsed_time < timeout() && spawned_solutions++ < properties().get<size_t>("max_solutions") - 1) {
+		// Randomize pose using specified dimension samplers.
+		// RPY dimensions are using 0.0 as seed, the randomized rotation is multiplied to the target orientation.
+		double rand_roll = 0.0, rand_pitch = 0.0, rand_yaw = 0.0;
+		for (const auto& pose_dim_sampler : pose_dimension_samplers_) {
+			switch (pose_dim_sampler.first) {
+				case X:
+					target_pose.pose.position.x = pose_dim_sampler.second(seed_pose.pose.position.x);
+					break;
+				case Y:
+					target_pose.pose.position.y = pose_dim_sampler.second(seed_pose.pose.position.y);
+					break;
+				case Z:
+					target_pose.pose.position.z = pose_dim_sampler.second(seed_pose.pose.position.z);
+					break;
+				case ROLL:
```

Comment:
Note that uniform sampling of Euler angles gives you very non-uniform sampling of orientations (you oversample at the "poles"). See https://github.com/ompl/ompl/blob/master/src/ompl/util/src/RandomNumbers.cpp#L279-L285.

Reply:
Never mind. I see you can you still do what the code does I just referred to. You would have three stages, one each for RPY. Roll and yaw can be sampled uniformly. You'd just have to pass a special sampler for pitch that computes arccos of a uniformly sampled number with some offset.

Reply:
We don't have control over the order of sampling yet (it's a map) and the already sampled RPY arguments are not passed to the sampler yet. So, I think, your original argument is still valid @mamoll.

Reply:
The stage now supports specifying the sampling order